tools/clang/tools/dxcvalidator/dxcvalidator.cpp

```diff
@@ -180,18 +217,39 @@ uint32_t hlsl::validateWithOptDebugModule(
       ULONG cbWritten;
       DiagStream->Write(msg.c_str(), msg.size(), &cbWritten);
     }
-    // Assemble the result object.
-    CComPtr<IDxcBlob> pDiagBlob;
-    hr = DiagStream.QueryInterface(&pDiagBlob);
-    DXASSERT_NOMSG(SUCCEEDED(hr));
-    hr = DxcResult::Create(
-        validationStatus, DXC_OUT_NONE,
-        {DxcOutputObject::ErrorOutput(
-            CP_UTF8, // TODO Support DefaultTextCodePage
-            (LPCSTR)pDiagBlob->GetBufferPointer(), pDiagBlob->GetBufferSize())},
-        Result);
-    if (FAILED(hr))
-      throw hlsl::Exception(hr);
+    if (Flags & (DxcValidatorFlags_ModuleOnly | DxcValidatorFlags_SkipHash)) {
+      // If we are validating a module only, we don't want to return the
+      // diagnostics stream.
+      CComPtr<IDxcBlob> pDiagBlob;
+      hr = DiagStream.QueryInterface(&pDiagBlob);
+      DXASSERT_NOMSG(SUCCEEDED(hr));
+      hr = DxcResult::Create(validationStatus, DXC_OUT_NONE,
+                             {DxcOutputObject::ErrorOutput(
+                                 CP_UTF8, // TODO Support DefaultTextCodePage
+                                 (LPCSTR)pDiagBlob->GetBufferPointer(),
+                                 pDiagBlob->GetBufferSize())},
+                             Result);
+      if (FAILED(hr))
+        throw hlsl::Exception(hr);
+    } else {
+      CComPtr<IDxcBlob> HashedBlob;
+      // Assemble the result object.
+      CComPtr<IDxcBlob> DiagBlob;
+      CComPtr<IDxcBlobEncoding> DiagBlobEnconding;
+      hr = DiagStream.QueryInterface(&DiagBlob);
+      DXASSERT_NOMSG(SUCCEEDED(hr));
+      hr = DxcCreateBlobWithEncodingSet(DiagBlob, CP_UTF8, &DiagBlobEnconding);
+      if (FAILED(hr))
+        throw hlsl::Exception(hr);
+      HashAndUpdateOrCopy(Flags, Shader, &HashedBlob);
+      hr = DxcResult::Create(
+          validationStatus, DXC_OUT_OBJECT,
+          {DxcOutputObject::DataOutput(DXC_OUT_OBJECT, HashedBlob),
+           DxcOutputObject::DataOutput(DXC_OUT_ERRORS, DiagBlobEnconding)},
+          Result);
+      if (FAILED(hr))
+        throw hlsl::Exception(hr);
+    }
```

Comment:
I think we can test this behavior in a CompilerTest

Reply:
Added test to ValidationTest.